src/Microsoft.Extensions.Primitives/StringValues.cs

```diff
@@ -27,8 +27,71 @@ public StringValues(string value)
 
         public StringValues(string[] values)
         {
-            _value = null;
-            _values = values;
+            if (values == null || values.Length == 0)
+            {
+                // Set null and empty arrays to null
+                _value = null;  // Both null is required for 
+                _values = null; // IsNull property to work correctly
+            }
+            else if (values.Length == 1)
+            {
+                // Set single item arrays to the string value
+                _values = null;
+                _value = values[0];
+            }
+            else
+            {
+                // Remove any nulls from the array else counts will be off
```

Comment:
This is beyond StringValues' purview to modify the input like this. And as far as we know this is a rare corner case so why put the extra effort into it? It's not even worth writing the tests for. None of your other changes in this PR depend on this filter, correct?

Reply:
Its an edge case with consequences.

The `string[]` ctor is the only way `null` can appear with a count
```csharp
new StringValues(new []{(string)null, (string)null, (string)null })
```
Will report `.Count == 3`

However, it will also output zero headers; so is it really 3 headers?

As `null` can get into `StringValues` values every header value has to be null checked in Kestrel to see if it is null or not before outputting it.

This means the previous will report `.Count == 0` and all the null checks can be removed from Kestrel for header output.

Reply:
Common case is there aren't nulls and the `string[]` ctor is not used; so its pushing to cost for potential null values coming from that .ctor into the .ctor; rather than having guards spread everywhere.

Reply:
The constructor is called from an implicit converter, it shouldn't be doing non-trivial transformations. StringValues also only wraps the array so you can't prevent those values from becoming null after the constructor.

Since Kestrel needs the values to not be null then it should do a one time filter, possibly via a helper method provided by StringValues like WithoutNulls or GetNonNulls.

Reply:
> Since Kestrel needs the values to not be null then it should do a one time filter, possibly via a helper method provided by StringValues like WithoutNulls or GetNonNulls.

That wouldn't achieve any real benefit over what exists now (and likely would be more expensive)

Reverted